raiden-ts/src/services/epics.ts

```diff
@@ -429,15 +442,133 @@ export const monitorUdcBalanceEpic = (
   {}: Observable<RaidenAction>,
   {}: Observable<RaidenState>,
   { address, latest$, userDepositContract }: RaidenEpicDeps,
-): Observable<udcDeposited> =>
+): Observable<udcDeposit.success> =>
   latest$.pipe(
     pluckDistinct('state', 'blockNumber'),
     // it's seems ugly to call on each block, but UserDepositContract doesn't expose deposits as
     // events, and ethers actually do that to monitor token balances, so it's equivalent
     exhaustMap(() => userDepositContract.functions.effectiveBalance(address) as Promise<UInt<32>>),
-    distinctUntilChanged((x, y) => y.eq(x)),
-    map(udcDeposited),
+    withLatestFrom(latest$),
+    filter(([balance, { udcBalance }]) => !udcBalance.eq(balance)),
+    map(([balance]) => udcDeposit.success(undefined, { totalDeposit: balance })),
+  );
+
+function makeUdcDeposit$(
+  [tokenContract, userDepositContract]: [HumanStandardToken, UserDeposit],
+  [sender, address]: [Address, Address],
+  [deposit, totalDeposit]: [UInt<32>, UInt<32>],
+  { pollingInterval, minimumAllowance }: RaidenConfig,
+  { log }: Pick<RaidenEpicDeps, 'log'>,
+) {
+  return defer(() =>
+    Promise.all([
+      tokenContract.functions.balanceOf(sender),
+      tokenContract.functions.allowance(sender, userDepositContract.address),
+      userDepositContract.functions.effectiveBalance(address),
+    ]),
+  ).pipe(
+    mergeMap(([balance, allowance, deposited]) => {
+      assert(deposited.add(deposit).eq(totalDeposit), [
+        ErrorCodes.UDC_DEPOSIT_OUTDATED,
+        { requested: totalDeposit.toString(), current: deposited.toString() },
+      ]);
+      assert(balance.gte(deposit), [
+        ErrorCodes.RDN_INSUFFICIENT_BALANCE,
+        { current: balance.toString(), required: deposit.toString() },
+      ]);
+
+      if (allowance.gte(deposit)) return of(true); // if allowance already enough
+
+      // secure ERC20 tokens require changing allowance only from or to Zero
+      // see https://github.com/raiden-network/light-client/issues/2010
+      let resetAllowance$: Observable<true> = of(true);
+      if (!allowance.isZero())
+        resetAllowance$ = defer(() =>
+          tokenContract.functions.approve(userDepositContract.address, 0),
+        ).pipe(
+          assertTx('approve', ErrorCodes.RDN_APPROVE_TRANSACTION_FAILED, { log }),
+          mapTo(true),
+        );
```

Comment:
Already thought about it. This is hardly splittable, since these return observable pipelines with different operations (`deposit` vs `setTotalDeposit`). These should be the only 2 types of contracts we approve for deposit, but I'll create an issue to evaluate if this is needed to be split/refactored.

Reply:
I've split `approveIfNeeded$`, should be enough, what's left is small and specific enough to stay its own functions.